### ISSUE TITLE
special case `typing.get_args` return type

### ISSUE THREAD

**Issue by KotlinIsland** [KotlinIsland/basedmypy#599]
```py
reveal_type(get_args(Literal[1, 2]))  # tuple[Any, ...]  expected, tuple[1, 2]
```

**Comment by jorenham** [comment]
The `typing._LiteralSpecialForm` special-type-expression-statement-alias-thingy isn't generic, so the type information is lost. 

So if it would've been something like `typing._LiteralSpecialForm[*Ts]`, it would be a piece of cake to add an overload like `get_args(thing: _LiteralSpecialForm[*Ts], /) -> tuple[*Ts]`, but alas...

**Comment by KotlinIsland** [comment]
yeah, it would be good to fix that as well. I think this would be complicated by the removal of duplicates and flattening etc, so probably not so simple

**Comment by jorenham** [comment]
> yeah, it would be good to fix that as well. I think this would be complicated by the removal of duplicates and flattening etc, so probably not so simple

I suppose that flattening should probably be sorted out in the `typing._LiteralSpecialForm.__new__`

But (unfortunately) `get_args` doesn't do much flattening in practice; it only flattens if directly nested, and isn't aware of e.g. `Literal[0] | Literal[1]` or anything that involves a `TypeAliasType` (i.e. almost everything, especially in Python 3.12):

```pycon
>>> from typing import Literal, TypeAlias, get_args
>>> Falsy: TypeAlias = Literal[None] | Literal[False, 0] | Literal['', b'']
>>> get_args(Falsy)
(typing.Literal[None], typing.Literal[False, 0], typing.Literal['', b''])
```

in case that wasn't infuriating enough, take a look at this one:

```pycon
>>> from typing import get_args
>>> type StringLike = str | bytes
>>> get_args(StringLike)
()
```

So you should probably don't even bother with `typing.get_args` at all, and use the (based)  [`optype.inspect.get_args()`](https://github.com/jorenham/optype?tab=readme-ov-file#optypeinspect) instead 🤷🏻.
However, it's still impossible to properly type it, because of what said in my previous comment.
So even though [optype's `get_args` annotations](https://github.com/jorenham/optype/blob/1228112deef5dbbd590b1207173f5d7cd41e9159/optype/inspect.py#L210-L244) are better than the [typeshed ones of  `typing.get_args`](https://github.com/python/typeshed/blob/089953ed841a201137f81dbd276767d2af6b3072/stdlib/typing.pyi#L873), they're still mostly useless.

BTW, I attempted to solve this a wihle back with a `Protocol`-based approach (i.e. by binding to the type of e.g. `__args__` of `Literal`, or the `__value__` of a `TypeAliasType`). But that also doesn't work, as those attributes aren't properly typed at all, because obviously there's no need to annotate the typing library itself..................@#!@#!#!!#!@!!!

... *ahem*

anyway, to make a short story long, I suppose that the only way to work around this mess, would be indeed like you suggested; fix it in the static type-checkers.

**Comment by KotlinIsland** [comment]
yeah, annoyingly `opttype`s work properly ones would require different special casing to `typing`s ones, i wonder if a simply mypy plugin could handle any of this?

**Comment by jorenham** [comment]
A mypy plugin could probably probably work. But I'm not much of a fan of mypy plugins, as they only widen the gap between type-checkers, and could be misused as an excuse to not properly fix it at the core.

**Comment by jorenham** [comment]
Perhaps we could attempt a `typeshed` PR  😅?

**Comment by KotlinIsland** [comment]
feel free to try and fix in basedmypys censored typwshed

**Comment by jorenham** [comment]
> feel free to try and fix in basedmypys vendored typeshed

That could actually be a rather clean solution to this issue, now that I think about it 🤔

**Comment by jorenham** [comment]
Considering the [`get_args` implementation](https://github.com/python/cpython/blob/v3.13.0rc2/Lib/typing.py#L2553-L2576), this'd require generic type annotations for the following types

- [`types.UnionType`](https://github.com/KotlinIsland/basedmypy/blob/4c2c2e5000d53f89d59fa0b208be793b71e9c2fe/mypy/typeshed/stdlib/types.pyi#L613-L620)
- [`types.GenericAlias`](https://github.com/KotlinIsland/basedmypy/blob/4c2c2e5000d53f89d59fa0b208be793b71e9c2fe/mypy/typeshed/stdlib/types.pyi#L581-L602) (**warning:** typing gore! Viewer discretion is advised.)
- `typing._AnnotatedAlias` (doesn't exist in typeshed)
- `typing._GenericAlias`  (doesn't exist in typeshed)
- [`typing._SpecialForm`](https://github.com/KotlinIsland/basedmypy/blob/4c2c2e5000d53f89d59fa0b208be793b71e9c2fe/mypy/typeshed/stdlib/typing.pyi#L193-L198) (a leprous sloth wrote this, probably)

Oh yea and the `typing_extensions` also should be updated in the same way.

---

>>> feel free to try and fix in basedmypys vendored typeshed

I guess it's doable, but it'd probably take me day or two. And because I'm currently working on something like 10 different projects simultaneously, it'll probably be a while until I get around to it 😅.
So if you want to solve this quickly, you might wanna place your bets on someone else who's willing to do this; I wouldn't mind help reviewing 🤷🏻
